app/src/main/java/com/example/ohthmhyh/database/DatabaseAdapter.java

```diff
@@ -123,6 +155,314 @@ public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
+            Log.d("DatabaseAdapter'", String.valueOf(habitList.size()));
+            Log.d("DatabaseAdapter'", String.valueOf(habitEventList.size()));
+            Log.d("DatabaseAdapter'", user.getUsername());
+            callback.onLoaded();
+        } else {
+            UID = FirebaseAuth.getInstance().getCurrentUser().getUid();
+            DatabaseAdapter.this.pullHabitEventList(new OnLoadedListener() {
+                @Override
+                public void onLoaded() {
+                    DatabaseAdapter.this.pullHabitList(new OnLoadedListener() {
+                        @Override
+                        public void onLoaded() {
+                            DatabaseAdapter.this.pullUser(new OnLoadedListener() {
+                                @Override
+                                public void onLoaded() {
+                                    Log.d("DatabaseAdapter", String.valueOf(habitList.size()));
+                                    Log.d("DatabaseAdapter", String.valueOf(habitEventList.size()));
+                                    Log.d("DatabaseAdapter", user.getUsername());
+                                    callback.onLoaded();
+                                }
+                            });
+                        }
+                    });
+                }
+            });
+        }
+    }
+
+
+    public void createDataForNewUser(String username) {
+        habitList = new HabitList();
+        habitEventList = new HabitEventList();
+        user = new User(username);
+
+        pushHabits(habitList);
+        pushHabitEvents(habitEventList);
+        pushUser(user);
+    }
+
+
+    public void pullHabitList(DatabaseAdapter.OnLoadedListener callback) {
+        DocumentReference documentReference = db.collection("Habits").document(UID);
+        documentReference.get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
+            @Override
+            public void onSuccess(DocumentSnapshot documentSnapshot) {
+                habitList = documentSnapshot.toObject(HabitList.class);
+                Log.d("DatabaseAdapter", "HabitList: " + String.valueOf(habitList.size()));
+                callback.onLoaded();
+            }
+        });
+    }
+
+    public void pullHabitsForUser(String userUID, DatabaseAdapter.OnLoadedHabitsListener callback) {
+        DocumentReference documentReference = db.collection("Habits").document(userUID);
+        documentReference.get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
+            @Override
+            public void onSuccess(DocumentSnapshot documentSnapshot) {
+                ArrayList<Habit> habits = new ArrayList<>();
+                HabitList habitList = documentSnapshot.toObject(HabitList.class);
+
+                for (int i = 0; i < habitList.size(); i++) {
+                    habits.add(habitList.getHabit(i));
+                }
+
+                callback.onLoadedHabits(habits);
```

Comment:
Yeah I was thinking about this, but then I thought it may be best to have a separate copy of the Habit ArrayList? This way the HabitList does not have to be kept in memory just for its HabitList to be used elsewhere? Let me know your thoughts.

Reply:
Makes sense to me! I thought this might be why you were doing that, but I wanted to make sure.

Reply:
Makes sense to me! I thought this might be why you were doing this, but I wanted to make sure